File: time_varying_pdnn.m
% time_varying_pdnn.m
% MATLAB translation of time_varying_pdnn.py
%
% Time-Varying Certified Primal-Dual Neural Network (PDNN) solver
% for time-varying quadratic programming (TVQP).
%
% Implements the projected gradient PDNN iteration:
%   y_candidate = y - alpha * gamma * (H*y + p)
%   y_new       = P_Omega(y_candidate)   % box projection via min/max
%
% KKT vector:  y = [x (3x1); lambda (2x1); mu (1x1)]
%   x      - primal decision variables (box-constrained)
%   lambda - equality constraint dual variables (unbounded)
%   mu     - inequality constraint dual variable (non-negative)
%
% Reference: 2026 IEEE TAC - EIQP (Section VI TVQP example, Fig. 3)

clear; clc;

%% Parameters
gamma          = 5000;   % PDNN scaling factor (Fig. 3(c))
epsilon        = 1e-5;   % Convergence tolerance for inner loop
max_inner_iter = 50;     % Maximum inner PDNN iterations per time step
omega          = 1.0;    % Frequency parameter (Fig. 3 setup)
dt             = 0.001;  % Internal step size
UNBOUNDED_LIM  = 1e6;    % Effectively unbounded limit for dual variable bounds

T_END    = 10;
N_POINTS = 1000;

%% Time span and result storage
t_span    = linspace(0, T_END, N_POINTS);
y_current = ones(6, 1);           % Initial state: n=3 primal, m=2 eq., k=1 ineq.
results   = zeros(N_POINTS, 3);   % Each row stores primal x(t) at that time step

%% Main simulation loop
for idx = 1:N_POINTS
    t = t_span(idx);
    y_current = solve_step(t, y_current, gamma, epsilon, ...
                           max_inner_iter, dt, omega, UNBOUNDED_LIM);
    results(idx, :) = y_current(1:3)';
end

fprintf('Simulation complete for t=[0, 10]s.\n');
fprintf('Final Solution x(10): [%.6f, %.6f, %.6f]\n', results(end, :));

%% ============================= Local Functions =============================

function [W, q, J, d, A, b, xi_min, xi_max] = get_matrices(t, omega)
% GET_MATRICES  Build time-varying TVQP coefficient matrices at time t.
%
%   Inputs:
%     t     - current simulation time
%     omega - frequency parameter
%
%   Outputs:
%     W      - 3x3 symmetric positive definite cost matrix
%     q      - 3x1 linear cost vector
%     J      - 2x3 equality constraint matrix  (J*x = d)
%     d      - 2x1 equality constraint RHS
%     A      - 1x3 inequality constraint matrix (A*x <= b)
%     b      - 1x1 inequality constraint RHS
%     xi_min - 3x1 lower bounds on primal variables
%     xi_max - 3x1 upper bounds on primal variables

    w = omega;

    W = [2*cos(w*t)+22,   cos(w*t)-2,       3*sin(w*t)+6;
         cos(w*t)-2,      2*cos(2*w*t)+12,  sin(w*t);
         3*sin(w*t)+6,    sin(w*t),          cos(3*w*t)+8];

    q = [sin(3*w*t); cos(3*w*t); -cos(2*w*t)];

    J = [2*sin(4*w*t),  cos(w*t),      sin(w*t)+4;
         0.5*cos(w*t),  0.5*sin(w*t),  sin(2*w*t)];

    d = [sin(2*w*t); -cos(4*w*t)];

    A = [0.5*cos(w*t)+2,  sin(w*t)+1,  sin(4*w*t)+1];

    b = 1.5*cos(w*t)+8;

    xi_min = [cos(4*w*t)-6; sin(w*t)-6; sin(w*t+2)-6];
    xi_max = [cos(4*w*t)+6; sin(w*t)+6; sin(w*t+2)+6];
end


function y = solve_step(t, y_prev, gamma, epsilon, max_inner_iter, dt, omega, UNBOUNDED_LIM)
% SOLVE_STEP  One time step of the projected-gradient PDNN iteration.
%
%   Builds the 6x6 KKT matrix H and 6x1 vector p from the TVQP data,
%   then iterates the projected gradient update until convergence or
%   max_inner_iter is reached.
%
%   KKT structure (block partition [3 | 2 | 1]):
%     H = [ W    -J'    A' ]
%         [ J     0      0 ]
%         [-A     0      0 ]
%
%   Inputs:
%     t              - current time
%     y_prev         - 6x1 KKT vector from the previous time step
%     gamma          - PDNN scaling factor
%     epsilon        - convergence tolerance
%     max_inner_iter - maximum inner iterations
%     dt             - step size (alpha)
%     omega          - frequency parameter
%     UNBOUNDED_LIM  - bound value for effectively unconstrained variables
%
%   Output:
%     y - updated 6x1 KKT vector

    [W, q, J, d, A, b, xi_min, xi_max] = get_matrices(t, omega);

    % Assemble KKT matrix H (6x6)
    H = [W,    -J',          A';
         J,     zeros(2,2),  zeros(2,1);
        -A,     zeros(1,2),  zeros(1,1)];

    % Assemble KKT vector p (6x1)
    p = [q; -d; b];

    % Box-constraint bounds for the projection set Omega
    lb = [xi_min;                     -UNBOUNDED_LIM * ones(2,1); zeros(1,1)         ];
    ub = [xi_max;                      UNBOUNDED_LIM * ones(2,1); UNBOUNDED_LIM * ones(1,1)];

    alpha = dt;
    y = y_prev;

    for iter = 1:max_inner_iter
        y_candidate = y - alpha * gamma * (H * y + p);

        % Project onto box constraints (equivalent to np.clip)
        y_new = min(max(y_candidate, lb), ub);

        % Early termination if converged
        if norm(y_new - y) < epsilon
            y = y_new;
            return;
        end
        y = y_new;
    end
end

File: time_varying_pdnn.py
import numpy as np

# Represents an effectively unbounded limit for dual variable box constraints
_UNBOUNDED_LIMIT = 1e6


class TimeVaryingCertifiedPDNN:
    def __init__(self, gamma=1e4, epsilon=1e-5, max_inner_iter=50):
        self.gamma = gamma
        self.epsilon = epsilon
        self.max_inner_iter = max_inner_iter
        self.omega = 1.0  # From Fig. 3 setup
        self.dt = 0.001   # Time step for simulation

    def get_matrices(self, t):
        w = self.omega
        # Simplified reconstruction of the Section VI TVQP coefficients
        W = np.array([
            [2*np.cos(w*t)+22, np.cos(w*t)-2, 3*np.sin(w*t)+6],
            [np.cos(w*t)-2, 2*np.cos(2*w*t)+12, np.sin(w*t)],
            [3*np.sin(w*t)+6, np.sin(w*t), np.cos(3*w*t)+8]
        ])
        q = np.array([np.sin(3*w*t), np.cos(3*w*t), -np.cos(2*w*t)])
        J = np.array([[2*np.sin(4*w*t), np.cos(w*t), np.sin(w*t)+4],
                      [0.5*np.cos(w*t), 0.5*np.sin(w*t), np.sin(2*w*t)]])
        d = np.array([np.sin(2*w*t), -np.cos(4*w*t)])
        A = np.array([[0.5*np.cos(w*t)+2, np.sin(w*t)+1, np.sin(4*w*t)+1]])
        b = np.array([1.5*np.cos(w*t)+8])
        xi_min = np.array([np.cos(4*w*t)-6, np.sin(w*t)-6, np.sin(w*t+2)-6])
        xi_max = np.array([np.cos(4*w*t)+6, np.sin(w*t)+6, np.sin(w*t+2)+6])
        return W, q, J, d, A, b, xi_min, xi_max

    def solve_step(self, t, y_prev):
        W, q, J, d, A, b, xi_min, xi_max = self.get_matrices(t)

        # Build H and p as per Eq. 6 (KKT system)
        # W: 3x3, J: 2x3, A: 1x3
        # y = [x (3), lambda (2), mu (1)]
        # Bug 1 fixed: zero blocks use correct semantic dimensions
        H = np.block([
            [W,  -J.T,              A.T             ],
            [J,   np.zeros((2, 2)), np.zeros((2, 1))],
            [-A,  np.zeros((1, 2)), np.zeros((1, 1))]
        ])
        p = np.concatenate([q, -d, b])

        # Projection bounds for Set Omega
        lb = np.concatenate([xi_min, -_UNBOUNDED_LIMIT*np.ones(2), np.zeros(1)])
        ub = np.concatenate([xi_max,  _UNBOUNDED_LIMIT*np.ones(2), _UNBOUNDED_LIMIT*np.ones(1)])

        # Bug 2, 3, 4 fixed: standard projected gradient update using gamma as
        # scaling factor; the M = I + H.T preconditioner has been removed.
        # The PDNN iteration solves: dy/dt = gamma * (P_Omega(y - (Hy+p)) - y)
        alpha = self.dt  # Internal step size matches configured dt
        y = y_prev.copy()

        for _ in range(self.max_inner_iter):
            y_candidate = y - alpha * self.gamma * (H @ y + p)
            y_new = np.clip(y_candidate, lb, ub)

            # Bug 5 fixed: use epsilon for early termination
            if np.linalg.norm(y_new - y) < self.epsilon:
                y = y_new
                break
            y = y_new

        return y


# --- Main Simulation Loop ---
# gamma=5000 as per Fig 3(c); T_END and N_POINTS control the time grid
T_END = 10
N_POINTS = 1000  # At least 1000 points for sufficient time resolution

solver = TimeVaryingCertifiedPDNN(gamma=5000)

# Bug 6 fixed: use at least 1000 points for sufficient time resolution
t_span = np.linspace(0, T_END, N_POINTS)

y_current = np.ones(6)  # Initial state: n=3, m=2, k=1
results = []

for t in t_span:
    y_current = solver.solve_step(t, y_current)
    results.append(y_current[:3])  # Extract primal x(t)

print(f"Simulation complete for t=[0, 10]s.")
print(f"Final Solution x(10): {results[-1]}")
